### PR TITLE
[iov-lisk] Create Lisk codec

### LINKED ISSUE #117: Add tx codec support for lisk format

### PR DESCRIPTION
**Pull request by webmaster128**
Closes #117

- [x] refactor away BaseTx.timestamp
- [x] Check max memo length
- [x] Disallow the use of the same passphrase twice in a keyring entry

### CODE REVIEW

File: packages/iov-lisk/package.json [hunk 1]
```diff
@@ -0,0 +1,40 @@
+{
+  "name": "@iov/lisk",
+  "version": "0.5.4",
+  "description": "Transction Codec and Client to communicate with the Lisk blockchain",
```
Comment (Isabello) [inline packages/iov-lisk/package.json]: Transaction

**Comment by webmaster128** [comment]
In the new README.md you can get the idea of how to use those tools. I'd be happy for any feedback that does is not included in the checklist above.

**Comment by webmaster128** [comment]
@Isabello @ethanfrey this is done from my side. Next up: network connection

File: packages/iov-lisk/README.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+```ts
+import { liskCodec, LiskKeyringEntry } from "@iov/lisk";
+
+const liskTestnet = "da3ed6a45429278bac2666961289ca17ad86595d33b31037615d4b8e8f158bba" as ChainId;
+
+const entry = new LiskKeyringEntry();
+const mainIdentity = await entry.createIdentity("oxygen fall sure lava energy veteran enroll frown question detail include maximum");
+
+const recipientAddress = Encoding.toAscii("6076671634347365051L") as Address;
```
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-lisk/README.md]: It's the Testnet faucet address. Real devs will see our love letters there

https://testnet-faucet.lisk.io/
https://testnet-faucet.lisk.io/

File: packages/iov-lisk/README.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+This is how you use `liskCodec` and `LiskKeyringEntry` to generate send transactions
+for Lisk manually, i.e. without the help of @iov/core.
+
+The following example is made for use in @iov/cli and you may need to include some
+missing symbols if used in a different environment.
+
+```ts
+import { liskCodec, LiskKeyringEntry } from "@iov/lisk";
+
+const liskTestnet = "da3ed6a45429278bac2666961289ca17ad86595d33b31037615d4b8e8f158bba" as ChainId;
```
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-lisk/README.md]: The Lisk "nethash" is a hash of the genesis block. It least they say so. I was not yet able to reproduce it. In any case, it is a network identifier the community agrees on.

File: packages/iov-lisk/src/derivation.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import Long from "long";
+
+import { Ed25519, Ed25519Keypair, Sha256 } from "@iov/crypto";
+import { Encoding } from "@iov/encoding";
+
+export async function passphraseToKeypair(passphrase: string): Promise<Ed25519Keypair> {
```
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-lisk/src/derivation.ts]: No HD, this is just a custom way to make a keypair from a password string. If you really want to know it, here you go: https://medium.com/@simonwarta/lisk-dont-call-it-256-bit-security-851385c8072a

File: packages/iov-lisk/src/liskcodec.ts [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+import { serializeTransaction, transactionId } from "./serialization";
+
+export const liskCodec: TxCodec = {
+  /**
+   * Transaction serialization as in
+   * https://github.com/prolina-foundation/snapshot-validator/blob/35621c7/src/transaction.cpp#L36
+   */
+  bytesToSign: (unsigned: UnsignedTransaction, nonce: Nonce): SigningJob => {
+    const creationTimestamp = nonce.toNumber();
+    const creationDate = new ReadonlyDate(creationTimestamp * 1000);
```
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-lisk/src/liskcodec.ts]: The issue is that JavaScript's DateTime is based on milliseconds but I wanted to store plain unix timestamps with second resolution in the field. Having millisecond resolution in the nonce is possible but not elegant as we use second resolution in Lisk transaction timestamps. So yes, a lot of back and forth but proper data in the nonce field.

File: packages/iov-lisk/src/liskcodec.ts [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    switch (signed.transaction.kind) {
+      case TransactionKind.Send:
+        const timestamp = signed.primarySignature.nonce.toNumber();
+        const liskTimestamp = timestamp - 1464109200;
+        const id = transactionId(
+          signed.transaction,
+          new ReadonlyDate(timestamp * 1000),
+          signed.primarySignature,
+        );
+        const amount = Long.fromNumber(
```
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-lisk/src/liskcodec.ts]: I am impressed by the quality of your review! Yes I want, and at the point of 75 % of Lisk supply I need.

Did this change as part of the next PR (https://github.com/iov-one/iov-core/pull/358/commits/d83bfff9c08ad888866452a16a0513fc14e52279)

File: packages/iov-lisk/src/liskkeyringentry.ts [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+
+  public getIdentities(): ReadonlyArray<LocalIdentity> {
+    return this.identities;
+  }
+
+  public setLabel(label: string | undefined): void {
+    this.labelProducer.update(label);
+  }
+
+  public async createIdentity(passphrase?: string): Promise<LocalIdentity> {
```
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-lisk/src/liskkeyringentry.ts]: Indeed, we need more clarity where the random generation takes place. Either in the entry or outside.

File: packages/iov-lisk/src/liskcodec.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-lisk/src/liskcodec.ts]: oh, wait, that was the other way round. In this case, the issue still exists (at 75 % of Lisk supply in one transaction). Nice catch, will fix.

File: packages/iov-lisk/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Isabello) [reply packages/iov-lisk/README.md]: I decided to use it as the nethash when we had people accidentally double spend.

File: packages/iov-lisk/src/liskcodec.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (webmaster128) [reply packages/iov-lisk/src/liskcodec.ts]: I am on it. Doing it on another branch as this affects more than one places in the code.

File: packages/iov-lisk/src/liskcodec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ethanfrey) [reply packages/iov-lisk/src/liskcodec.ts]: Thanks for the explanation. Would be great to have a comment mention that I the struct

**Pull request merged**